eneyj/docs/SPECIFICATION

```diff
@@ -52,231 +35,280 @@ Idea: Make keys evaluable (not implemented)
+anyone to create new parsers and linearizers and thus introduce interesting
+usage patterns.
 
-*x8.7) The language can be Abstract, in which case IDs are used instead of
-      labels.
+=== Types ===
 
-(Linearizers are still confusing. This needs a better design doc.
-* It seems we need two different linearizers, and thus should have two
-  different keys.
-* linearizer(one, English) should become "positive_integer("1")", but
-  linearizer([one], English) should be "[one]")
+A type contains a list of keys which are usually the keys that objects of this
+type will have.
+But the true arbiter of whether an object is valid or not is the validation
+function that is part of the type.
+This allows to add generic types by having functions that return a type with
+custom keys and a custom validation function.
 
-*9) Parser / Constructor
+For example, if we want to have a list of strings as a type, we can call
+a function "list" that takes as a parameter a type and returns a type, and
+if we called it with "list(string)" it will return a type that has a validator
+that ensures that an object fulfills all the conditions of a list and
+additionally that each element of the list is of type string.
 
-x9.1) There is a single function, Z parse, that turns a string into a Z Object,
-      given a language.
+In order to allow for type checking to work, all types have a key that
+quotes the function call that created the type.
 
-x9.2) That function needs to be extended if new syntactic sugar is being added,
-      e.g. for understanding array notation using "[]".
+=== Function calls ===
 
-*9.3) Check what's up with constructions.
+A function call is evaluated by executing an implementation as chosen by the
+evaluator in the preprocessing step.
+All implementations should always evaluate to the same result so the chosen
+implementation should not matter, modulo run-time behaviour.
+Implementations are objects of their own.
 
-10) Type system
+=== Z14 Implementations ===
 
-10.1) Every Z1 Z Object has a Z1K1 type of type Z4.
+A function can have several implementations associated with it.
+Every implementation must strictly follow the the specification of the function.
+All implementations of a function must return the same results for the same
+input if they are functional.
 
-x10.2) The type sysem is strict.
+Implementations can be given as built-ins (we hope to keep the numbers low)
+and the evaluator in that case has to be able to evalute those natively,
+or in written in a programming language the evaluator supports, or implemented
+by composition of other functions.
 
-x10.3) There is no type hierarchy.
+=== Generic functions ===
 
-(*10.4) The Types have a number of fields used for dispatching functions.
-        Maybe that shouldn't be like this. We need a design doc for dispatching.)
+How do we represent head : [list(string)] -> string ?
 
-11) Errors
+head_type : [type] -> zobject function(list(zobject))
```

Comment:
I'm not sure quite what notation is being used here, this doesn't entirely made sense. Do you have a reference? Is this related to the notation in Benjamin Pierce's book?

Reply:
I removed that. That section was too drafty for now.